Add login component

AntonioLok opened the pull request:
# Description
Backend implementation to validate user login and send token to client was completed in a previous PR. This PR implements the front-end logic of the login component. The jwt is sent from the backend and is stored in local storage in the client browser. 

# Test 
- In your local repository of the [back-end component](https://github.com/AntonioLok/online-shopping-reactJS-Backend), install latest dep using `npm i` and start application `npm run start`.
- In your local repository of the [front-end component](https://github.com/AntonioLok/online-shopping-reactJS-Frontend), install latest dep using `npm i` and start application `npm run start`.
- (If you haven't created an account yet), do so first [here](http://localhost:8080/register). 
- Then login with the same credentials [here](http://localhost:8080/login).
- Test that the token have been stored by going to the console and send the following: `localStorage.getItem("token")`
# Screenshots
- Bad login request due to incorrect password or user not being registered:
![image](https://user-images.githubusercontent.com/14690224/61258280-cc5f7980-a742-11e9-936e-74b4e01e1a09.png)
- Success: should navigate to the home component and you should be able to get your token as below:
![image](https://user-images.githubusercontent.com/14690224/61258305-f3b64680-a742-11e9-9dc0-07b0db6a76c2.png)


# To do after
Add e2e/VR test for the sign-up and log-in component

howardhenry commented on src/components/app.js:
```diff
@@ -7,13 +7,14 @@ import NavHeader from './common/nav-header';
 const history = createHistory();
 const {
-  home, register, products, product,
+  home, register, products, product, logIn,
```
Login (noun) is one word. So you can change this to all lowercase: "login"

AntonioLok replied:
Ah okay. My bad, I thought register was also a verb and I wanted to follow the same pattern so I went with `Log In` as in like the actual verb to log in. But now I see the register is also a noun.... I will update it.

howardhenry commented on src/components/user/login/index.js:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    if (isLoginSuccessful) {
+      localStorage.setItem('token', data);
```
Logic to setting the auth token should be handled in the reducer, not in the component. Remember the `render` method is fired every time a prop or state value changes. That means, with the current login you'll be unnecessarily storing the token multiple times on each of those re-renders, which is an anti-pattern as it degrades the application's performance.

AntonioLok replied:
Ah you are right. I didnt think about that one. Sorry I haven't done any react in a while but I thought it only re-renders when the state is updated or if shouldComponentUpdate() returns true. I never knew that it re-renders when props changes... that's new for me.

AntonioLok replied:
On second thought... I do not see how I am storing the token "multiple times". As soon as I store the token in the localstorage, I am navigating to the home page which means that there will no be re-render right after I've stored the token. There will no be cases where the application stores the token again unless an user navigates to the login page again and re-login. I will update the code nevertheless

howardhenry commented on src/components/user/login/login-form.js:
```diff
@@ -0,0 +1,22 @@
+import React, { PureComponent } from 'react';
```
This component is identical to _ src/components/user/register/register-form.js_ except for the async validate functionality. The extra layer of abstraction from _src/components/common/form/index.js_ doesn't seem necessary as there is no logic in the render function.

I suggest moving any logic in the `export default` statement to the _src/components/common/form/index.js_ abstraction. Refer to the reduxForm API documentation to see what options you have to make this code more DRY. There is an attribute in the options object called `form` which can be passed in to uniquely identify each form (https://redux-form.com/8.2.2/docs/api/props.md/#-code-form-string-code-)

AntonioLok replied:
This is a really great suggestion. Thanks! I no longer need these 2 files. :)

howardhenry replied:
👍

howardhenry commented on src/store/actions/login.js:
```diff
@@ -0,0 +1,22 @@
+export const loginAPI = user => async (dispatch) => {
+  try {
+    const response = await axios({
+      url: `${API_BASE_URL}/users/login`,
```
I would strongly suggest the implementation of a middleware to handle your API requests. You have a lot of repeated code:
1. prepending `API_BASE_URL` to your endpoint
2. triggering `dispatch` on the success and failure states and passing the response data to the redux store

This does not allow you to scale or refactor any of this logic quickly across you application.

You can handle this in a separate PR. But it should be handled as soon as possible as it is impacting other design decisions in your application. Perhaps you could even pause this PR until the middleware refactor is complete.

AntonioLok replied:
I will open  a PR after this to handle it as to not block this PR. i have created a ticket https://github.com/AntonioLok/online-shopping-reactJS-Frontend/projects/1 and its my next to-do item

howardhenry commented on src/store/actions/login.js:
```diff
@@ -0,0 +1,22 @@
+  payload,
+});
+
+export const loginAPI = user => async (dispatch) => {
```
`loginAPI` isn't a clear description for a function. Function names should typically be verbs as they describe the logic that is contained within them. `loginAPI` however is a noun. 

Once you have the appropriate middleware, you actually shouldn't need this `loginAPI` function as it will be automatically handled.

I'll leave you to research the implementation of the middleware to leverage a more DRY structure.

AntonioLok replied:
Yup. I will mark it as resolved as it will no longer be needed once I implement the middleware API on my next PR.

howardhenry commented on src/components/user/login/index.js:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    if (isLoginSuccessful) {
+      localStorage.setItem('token', data);
```
"token" is a bit ambiguous. I would suggest renaming this "OS_AUTH_TOKEN" so that:
1. it's clear that it is used for authentication
2. It is namespaced with "OS" so it avoids potential conflicts with other libraries that may use the same token name

howardhenry commented on src/components/user/login/index.js:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    if (isLoginSuccessful) {
+      localStorage.setItem('token', data);
```
Any reason why you went with localStorage vs cookie vs session storage?

From our previous discussion, it is important to understand the pros, cons, limitations etc of each of these storage methods and ensure that you're using the correct one that achieves what you want.

AntonioLok replied:
local storage vs session storage, I think this one might be kinda self-explanatory but I just wanted to have to token be more persistent. This is just personal preference. As a user, if I am logged in to a page and I accidentally close that tab, when I open a new tab and navigate to the same page I would like to stay logged in. 

local storage vs cookie, while I think cookies are a better option, I wanted to keep the implementation more simple and local storage just achieve what I want for now. Once I look to expand my application, I will re-consider using cookies.

howardhenry replied:
👍

howardhenry commented on src/components/user/login/index.js:
```diff
@@ -0,0 +1,82 @@
+
+Login.propTypes = {
+  login: PropTypes.func.isRequired,
+  loginState: PropTypes.object.isRequired,
```
You should avoid using `PropTypes.object` and instead use `PropTypes.shape`. `PropTypes.object` does not tell what the sub-properties look like. So, as a developer, I cannot know that I should expect, `statusCode`, `error`, `message` etc.

`PropTypes` serve only one purpose; to document the expected properties that may be inputted to the component.

howardhenry commented on src/components/user/login/index.js:
```diff
@@ -0,0 +1,82 @@
+
+  renderFormError() {
+    const { loginState } = this.props;
+    const { message } = loginState;
```
What is this message? And, where does it come from?

Messages in your application should be defined in constants in your frontend and not be delivered from your backend. Having them in your frontend will allow you to:
1. interpolate variable data in the messages
2. handle accessibility concerns
3. handle i18n translations
4. optimize the payload from the backend
5. ensure the UI text is independent from the backend

Refactor this to have your error and success messages mapped to status codes from the API call payload.

AntonioLok replied:
The message is coming from the backend. I never knew this... I always thought the point of including the message in the payload of the request sent from the back-end to the front-end was so we could re-use that same message and display it on the front-end component. Good to know~

AntonioLok commented:
@howardhenry just FYI, I still haven't addressed all your comments yet. I've "resolved" some of the comments just to keep track of which ones I've addressed locally. Once I am done addressing all comments, I will tag you again!

howardhenry commented:
> @howardhenry just FYI, I still haven't addressed all your comments yet. I've "resolved" some of the comments just to keep track of which ones I've addressed locally. Once I am done addressing all comments, I will tag you again!

👍

AntonioLok commented:
@howardhenry I have addressed all your comments and updated my PR accordingly. When you get the chance, please review it :). Thanks!

howardhenry commented on src/stylesheets/components/_form.scss:
```diff
@@ -47,3 +47,7 @@ form {
   }
 }
+
+.alert-danger {
```
This style doesn't quite belong here. It is not within the `form` style block, which means it will be applied in any component that uses `.alert-danger`. If this style is intended to be applied everywhere, it should be more appropriately placed in it's own file _../alert.scss_

AntonioLok replied:
Yup, I wanted to have the css applied to the `.alert-danger` element only when its used for a form. However the alert-danger is being mounted outside of the form element, which means that there is no way I can have the css for `.alert-danger` nested within the form style block. I will take your advice and place it in its own file
